core/src/data/mbtilesDataSource.cpp

```diff
@@ -0,0 +1,506 @@
+CREATE INDEX IF NOT EXISTS map_grid_id ON map (grid_id);
+-- CREATE INDEX IF NOT EXISTS geocoder_type_index ON geocoder_data (type);
+-- CREATE UNIQUE INDEX IF NOT EXISTS geocoder_shard_index ON geocoder_data (type, shard);
+
+CREATE VIEW IF NOT EXISTS tiles AS
+    SELECT
+        map.zoom_level AS zoom_level,
+        map.tile_column AS tile_column,
+        map.tile_row AS tile_row,
+        images.tile_data AS tile_data
+    FROM map
+    JOIN images ON images.tile_id = map.tile_id;
+
+CREATE VIEW IF NOT EXISTS grids AS
+    SELECT
+        map.zoom_level AS zoom_level,
+        map.tile_column AS tile_column,
+        map.tile_row AS tile_row,
+        grid_utfgrid.grid_utfgrid AS grid
+    FROM map
+    JOIN grid_utfgrid ON grid_utfgrid.grid_id = map.grid_id;
+
+CREATE VIEW IF NOT EXISTS grid_data AS
+    SELECT
+        map.zoom_level AS zoom_level,
+        map.tile_column AS tile_column,
+        map.tile_row AS tile_row,
+        keymap.key_name AS key_name,
+        keymap.key_json AS key_json
+    FROM map
+    JOIN grid_key ON map.grid_id = grid_key.grid_id
+    JOIN keymap ON grid_key.key_name = keymap.key_name;
+COMMIT;)SQL_ESC";
+
+struct MBTilesQueries {
+    // SELECT statement from tiles view
+    SQLite::Statement getTileData;
+
+    // REPLACE INTO statement in map table
+    SQLite::Statement putMap;
+
+    // REPLACE INTO statement in images table
+    SQLite::Statement putImage;
+
+    MBTilesQueries(SQLite::Database& _db, bool _cache)
+        : getTileData(_db, "SELECT tile_data FROM tiles WHERE zoom_level = ? AND tile_column = ? AND tile_row = ?;"),
+          putMap(_db, _cache ? "REPLACE INTO map (zoom_level, tile_column, tile_row, tile_id) VALUES (?, ?, ?, ?);" : ";" ),
+          putImage(_db, _cache ? "REPLACE INTO images (tile_id, tile_data) VALUES (?, ?);" : ";") {}
+
+};
+
+MBTilesDataSource::MBTilesDataSource(std::string _name, std::string _path,
+                                     std::string _mime, bool _cache, bool _offlineFallback)
+    : m_name(_name),
+      m_path(_path),
+      m_mime(_mime),
+      m_cacheMode(_cache),
+      m_offlineMode(_offlineFallback) {
+
+    m_worker = std::make_unique<AsyncWorker>();
+
+    openMBTiles();
+}
+
+MBTilesDataSource::~MBTilesDataSource() {
+}
+
+bool MBTilesDataSource::loadTileData(std::shared_ptr<TileTask> _task, TileTaskCb _cb) {
+
+    if (m_offlineMode) {
+        if (_task->rawSource == this->level) {
+            // Try next source
+            _task->rawSource = next->level;
+        }
+
+        return loadNextSource(_task, _cb);
+    }
+
+    if (!m_db) { return false; }
+
+    if (_task->rawSource == this->level) {
+
+        m_worker->enqueue([this, _task, _cb](){
+            TileID tileId = _task->tileId();
+
+            auto& task = static_cast<BinaryTileTask&>(*_task);
+            task.rawTileData = std::make_shared<std::vector<char>>();
+
+            getTileData(tileId, *task.rawTileData);
```

Comment:
If we can't allocate an empty vector of bytes here then crashing seems like an appropriate response.